internal/oci/container.go

```diff
@@ -761,3 +763,54 @@ func (c *Container) RuntimePathForPlatform(r *runtimeOCI) string {
 	}
 	return c.runtimePath
 }
+
+// AddExecPID registers a PID associated with an exec session.
+// It is tracked so exec sessions can be cancelled when the container is being stopped.
+// If the PID is conmon, shouldKill should be false, as we should not call SIGKILL on conmon.
+// If it is an exec session, shouldKill should be true, as we can't guarantee the exec process
+// will have a SIGINT handler.
+func (c *Container) AddExecPID(pid int, shouldKill bool) error {
+	c.stopLock.Lock()
+	defer c.stopLock.Unlock()
+	if c.stopping {
+		return errors.New("cannot register an exec PID: container is stopping")
+	}
+	c.execPIDs[pid] = shouldKill
+	return nil
+}
+
+// DeleteExecPID is for deregistering a pid after it has exited.
+func (c *Container) DeleteExecPID(pid int) {
+	c.stopLock.Lock()
+	defer c.stopLock.Unlock()
+	delete(c.execPIDs, pid)
+}
+
+// KillExecPIDs loops through the saved execPIDs and sends a signal to them.
+// If shouldKill is true, the signal is SIGKILL. Otherwise, SIGINT.
+func (c *Container) KillExecPIDs() {
+	c.stopLock.Lock()
+	toKill := c.execPIDs
+	c.stopLock.Unlock()
+
+	for len(toKill) != 0 {
+		unkilled := map[int]bool{}
+		for pid, shouldKill := range toKill {
+			if pid == 0 {
+				// Just in case
+				continue
+			}
+			sig := syscall.SIGINT
+			if shouldKill {
```

Comment:
We could save the multiple passes using a single loop over a slice. First, send SIGINT, then sleep, and then follow up with the SIGKILL. This is what it boils down in the end, isn't it?

Reply:
no we should never SIGKILL conmon, the `shouldKill` bool is basically `isConmon`. I hear you on the loop. We could just go through once if you'd prefer, but I am worried the sigint may be lost... WDYT?

Reply:
No issue with the SIGINT vs SIGKILl for conmon. My comment was more to avoid having to loop more than once. I also wanted to avoid using a map in Go, as these things are just utterly broken at the moment, per:

- https://github.com/golang/go/issues/20135

This bothers me greatly every time I see a map in Go nowadays.

Reply:
> My comment was more to avoid having to loop more than once.

if the processes exit, then we won't loop more :) A concern could be busy looping though...

> I also wanted to avoid using a map in Go, as these things are just utterly broken at the moment, per:

I don't think we need to worry as much about this, as the map will only survive for the lifecycle of the container. Though, if you'd prefer, we could track the pids in separate slices or something, though that would make it more expensive to delete them...

Reply:
@haircommander, sounds good!